prediction_market_agent/agents/blockchain_coding_agent/functions.py

```diff
@@ -0,0 +1,139 @@
+from eth_typing import ChecksumAddress
+from prediction_market_agent_tooling.config import RPCConfig
+from prediction_market_agent_tooling.markets.omen.omen_contracts import (
+    OmenConditionalTokenContract,
+)
+from prediction_market_agent_tooling.tools.contract import ContractOnGnosisChain
+from prediction_market_agent_tooling.tools.tavily.tavily_models import TavilyResponse
+from prediction_market_agent_tooling.tools.tavily.tavily_search import (
+    tavily_search as tavily_search_pmat,
+)
+from web3 import Web3
+
+
+@tenacity.retry(stop=tenacity.stop_after_attempt(3), wait=tenacity.wait_fixed(1))
+def tavily_search(
+    query: str,
+) -> TavilyResponse:
+    return tavily_search_pmat(query=query)
+
+
+def fetch_read_methods_from_blockscout(contract_address: str) -> t.Any:
+    w3 = OmenConditionalTokenContract().get_web3()
+    if not is_contract(w3, Web3.to_checksum_address(contract_address)):
+        raise ValueError(f"{contract_address=} is not a contract on Gnosis Chain.")
+    read_not_proxy = fetch_read_methods(contract_address)
+    read_proxy = fetch_read_methods_proxy(contract_address)
+    # ToDo - Fetch write methods
+    return read_not_proxy + read_proxy
+
+
+def fetch_read_methods(contract_address: str) -> list[dict[str, t.Any]]:
+    url = f"https://gnosis.blockscout.com/api/v2/smart-contracts/{contract_address}/methods-read?is_custom_abi=false"
+    r = requests.get(url)
+    data: list[dict[str, t.Any]] = r.json()
+    return data
+
+
+def is_contract(web3: Web3, contract_address: ChecksumAddress) -> bool:
+    return bool(web3.eth.get_code(contract_address))
+
+
+def fetch_read_methods_proxy(contract_address: str) -> list[dict[str, t.Any]]:
+    url = f"https://gnosis.blockscout.com/api/v2/smart-contracts/{contract_address}/methods-read-proxy?is_custom_abi=false"
+    r = requests.get(url)
+    data: list[dict[str, t.Any]] = r.json()
+    return data
+
+
+def get_rpc_endpoint() -> str:
+    return RPCConfig().gnosis_rpc_url
+
+
+def checksum_address(address: str) -> ChecksumAddress:
+    from web3 import Web3
+
+    return Web3.to_checksum_address(address)
+
+
+def execute_read_function(
+    contract_address: str,
+    abi: str,
+    function_name: str,
+    function_parameters: List[str] = [],
+) -> t.Any:
+    """
+    Purpose:
+        Executes a read function on a smart contract using the specified contract address, ABI, function name, and function parameters.
+
+    Args:
+        contract_address (str): The address of the smart contract on which to execute the function.
+        abi (str): The ABI (Application Binary Interface) of the smart contract.
+        function_name (str): The name of the function to execute on the smart contract.
+        function_parameters (list): A list of parameters to pass to the function.
+
+    Returns:
+        Any: The result of calling the specified function on the smart contract.
+
+    """
+    from prediction_market_agent_tooling.tools.contract import abi_field_validator
```

Comment:
Tavily is imported at top of the file, but this is imported within the function. Is there any reason for that?